plugins/ccgx/fu-ccgx-hpi-device.c

```diff
@@ -562,41 +907,159 @@ fu_ccgx_hpi_device_prepare_firmware (FuDevice *device,
 				     GError **error)
 {
 	FuCcgxHpiDevice *self = FU_CCGX_HPI_DEVICE (device);
-	g_autoptr(FuFirmware) firmware = fu_ccgx_cyacd_firmware_new ();
-	g_autoptr(GPtrArray) images = NULL;
+	FWMode fw_mode;
+	guint16 fw_app_type;
+	guint16 fw_silicon_id;
+	g_autoptr(FuFirmware) firmware = fu_ccgx_firmware_new ();
 
 	/* parse all images */
 	if (!fu_firmware_parse (firmware, fw, flags, error))
 		return NULL;
 
-	/* check the silicon ID of all images */
-	images = fu_firmware_get_images (firmware);
-	for (guint i = 0; i < images->len; i++) {
-		FuFirmwareImage *img = g_ptr_array_index (images, i);
-		FuCcgxCyacdFirmwareImage *img_ccgx = FU_CCGX_CYACD_FIRMWARE_IMAGE (img);
-		guint16 fw_app_type = fu_ccgx_cyacd_firmware_image_get_app_type (img_ccgx);
-		guint16 fw_silicon_id = fu_ccgx_cyacd_firmware_image_get_silicon_id (img_ccgx);
-		if (fw_silicon_id != self->silicon_id) {
-			g_set_error (error,
-				     FWUPD_ERROR,
-				     FWUPD_ERROR_NOT_SUPPORTED,
-				     "silicon id mismatch on image %u, "
-				     "expected 0x%x, got 0x%x",
-				     i, self->silicon_id, fw_silicon_id);
-			return NULL;
-		}
-		if (fw_app_type != self->fw_app_type) {
-			g_set_error (error,
+	/* check the silicon ID */
+	fw_silicon_id = fu_ccgx_firmware_get_silicon_id (FU_CCGX_FIRMWARE (firmware));
+	if (fw_silicon_id != self->silicon_id) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "silicon id mismatch, expected 0x%x, got 0x%x",
+			     self->silicon_id, fw_silicon_id);
+		return NULL;
+	}
+	fw_app_type = fu_ccgx_firmware_get_app_type (FU_CCGX_FIRMWARE (firmware));
+	if (fw_app_type != self->fw_app_type) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "app type mismatch, expected 0x%x, got 0x%x",
+			     self->fw_app_type, fw_app_type);
+		return NULL;
+	}
+	fw_mode = fu_ccgx_firmware_get_fw_mode (FU_CCGX_FIRMWARE (firmware));
+	if (fw_mode != fu_ccgx_fw_mode_get_alternate (self->fw_mode)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "FWMode mismatch, expected %s, got %s",
+			     fu_ccgx_fw_mode_to_string (fu_ccgx_fw_mode_get_alternate (self->fw_mode)),
+			     fu_ccgx_fw_mode_to_string (fw_mode));
+		return NULL;
+	}
+	return g_steal_pointer (&firmware);
+}
+
+static gboolean
+fu_ccgx_hpi_get_metadata_offset (FuCcgxHpiDevice *self,
+				 FWMode fw_mode,
+				 guint32 *addr,
+				 guint32 *offset,
+				 GError **error)
+{
+	guint32 addr_max;
+
+	/* sanity check */
+	if (self->flash_row_size == 0x0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "unset support row size: 0x%x",
+			     self->flash_row_size);
+		return FALSE;
+	}
```

Comment:
do you perhaps want another check comparing flash row size and flash size?

Reply:
Good idea, added.